lib/config/flat-config-schema.js

```diff
@@ -0,0 +1,362 @@
+/**
+ * @fileoverview Flat config schema
+ * @author Nicholas C. Zakas
+ */
+
+"use strict";
+
+//-----------------------------------------------------------------------------
+// Type Definitions
+//-----------------------------------------------------------------------------
+
+/**
+ * @typedef ObjectPropertySchema
+ * @property {Function|string} merge The function or name of the function to call
+ *      to merge multiple objects with this property.
+ * @property {Function|string} validate The function or name of the function to call
+ *      to validate the value of this property.
+ */
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+const ruleSeverities = new Map([
+    [0, 0], ["off", 0],
+    [1, 1], ["warn", 1],
+    [2, 2], ["error", 2]
+]);
+
+const globalVariablesValues = new Set([
+    true, "writable", "writeable",
+    false, "readonly",
+    "off"
+]);
+
+//-----------------------------------------------------------------------------
+// Assertions
+//-----------------------------------------------------------------------------
+
+/**
+ * Validates that a value is a valid rule options entry.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't a valid rule options.
+ */
+function assertIsRuleOptions(value) {
+
+    if (typeof value !== "string" && typeof value !== "number" && !Array.isArray(value)) {
+        throw new TypeError("Expected a string, number, or array.");
+    }
+}
+
+/**
+ * Validates that a value is valid rule severity.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't a valid rule severity.
+ */
+function assertIsRuleSeverity(value) {
+    const severity = typeof value === "string"
+        ? ruleSeverities.get(value.toLowerCase())
+        : ruleSeverities.get(value);
+
+    if (typeof severity === "undefined") {
+        throw new TypeError("Expected severity of \"off\", 0, \"warn\", 1, \"error\", or 2.");
+    }
+}
+
+
+/**
+ * Validates that a given string is the form pluginName/objectName.
+ * @param {string} value The string to check.
+ * @returns {void}
+ * @throws {TypeError} If the string isn't in the correct format.
+ */
+function assertIsPluginMemberName(value) {
+    if (!/[@a-z0-9-_$]+\/(?:[a-z0-9-_$]+)+/iu.test(value)) {
```

Comment:
The second `+` in `(?:[a-z0-9-_$]+)+` looks unnecessary (and might cause catastrophic backtracking if we add `$` anchor or anything else after `+` in the pattern), did you mean something else?

Reply:
Ooh nice catch. Yeah, the intent was to allow `foo/bar/baz` to map what we have in eslintrc, but clearly I messed that pattern up.